numpy/fft/_pocketfft_umath.c

```diff
@@ -0,0 +1,354 @@
+    npy_intp n_outer = dimensions[0];
+    npy_intp si = steps[0], sf = steps[1], so = steps[2];
+    npy_intp nin = dimensions[1], nout = dimensions[2];
+    npy_intp isi = steps[3], iso = steps[4];
+    int (*cfft_function)(cfft_plan, double *, double) = func;
+    npy_intp npts = nout;
+    char *buffer = NULL;
+    int no_mem = 1;
+    npy_bool inplace = 0;
+
+    cfft_plan plan = make_cfft_plan(npts);
+    if (plan == NULL) {
+        goto fail;
+    }
+    if (iso != sizeof(npy_cdouble)) {
+        buffer = malloc(npts * sizeof(npy_cdouble));
+        if (buffer == NULL) {
+            goto fail;
+        }
+    }
+    else {
+        inplace = (ip == op && nin == nout && si == so && isi == iso);
+    }
+
+    for (npy_intp i = 0; i < n_outer; i++, ip += si, fp += sf, op += so) {
+        double fct = *(double *)fp;
+        char *op_buf = buffer == NULL? op : buffer;
+        if (!inplace) {  /* no copy needed if in-place already */
+            copy_data(ip, isi, nin, op_buf, sizeof(npy_cdouble), npts, sizeof(npy_cdouble));
+        }
+        if ((no_mem = cfft_function(plan, (double *)op_buf, fct)) != 0) {
+            break;
+        }
+        if (op_buf != op) {
+            copy_data(op_buf, sizeof(npy_cdouble), npts, op, iso, npts, sizeof(npy_cdouble));
+        }
+    }
+  fail:
+    if (buffer != NULL) {
+        free(buffer);
+    }
+    if (plan != NULL) {
+        destroy_cfft_plan(plan);
+    }
+    if (no_mem) {
+        /* TODO: Requires use of new ufunc API to indicate error return */
+        NPY_ALLOW_C_API_DEF
+        NPY_ALLOW_C_API;
+        PyErr_NoMemory();
+        NPY_DISABLE_C_API;
+    }
+    return;
+}
+
+
+
+static void
+rfft_impl(char **args, npy_intp const *dimensions, npy_intp const *steps, void *func,
+    npy_intp npts)
+{
+    char *ip = args[0], *fp = args[1], *op = args[2];
+    npy_intp n_outer = dimensions[0];
+    npy_intp si = steps[0], sf = steps[1], so = steps[2];
+    npy_intp nin = dimensions[1], nout = dimensions[2];
+    npy_intp isi = steps[3], iso = steps[4];
+    char *buffer = NULL;
+    int no_mem = 1;
+
+    rfft_plan plan = make_rfft_plan(npts);
+    if (plan == NULL) {
+        goto fail;
+    }
+    if (iso != sizeof(npy_cdouble)){
+        buffer = malloc(nout * sizeof(npy_cdouble));
+        if (buffer == NULL) {
+            goto fail;
+        }
+    }
+
+    for (npy_intp i = 0; i < n_outer; i++, ip += si, fp += sf, op += so) {
+        double fct = *(double *)fp;
+        char *op_buf = buffer == NULL? op : buffer;
+        double *op_double = (double *)op_buf;
+        /* Copy dat to buffer in starting at position 1, as expected for FFTpack order */
+        copy_data(ip, isi, nin,
+                  (char *)&op_double[1], sizeof(npy_double), nout*2 - 1, sizeof(npy_double));
+        if ((no_mem = rfft_forward(plan, &op_double[1], fct)) != 0) {
+            break;
+        }
+        op_double[0] = op_double[1];
+        op_double[1] = 0.;
+        if (op_buf != op) {
```

Comment:
Maybe using `buffer != NULL` (and below) is clearer.  Maybe not...

Reply:
I had that at some point, but didn't like that then the copy was from `op_buf`. I now changed the name to `op_or_buff` and do `if (op_or_buff == buff)` which I think is clearer.